File: contract-tests/runners/java/Phase0Runner.java
import com.google.protobuf.Any;
import io.lifechronicle.events.v1.AppForeground;
import io.lifechronicle.events.v1.EventEnvelope;
import io.lifechronicle.events.v1.Origin;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Properties;

/**
 * Independent Java implementation of the Phase 0 golden-vector framing rules.
 *
 * <p>The expected.* properties are deliberately never read. The runner derives
 * every emitted digest and frame from the input fields, then uses the generated
 * Java Protobuf bindings for the decode half of the contract.
 */
public final class Phase0Runner {
  private static final String VECTOR_NAME = "phase0-v1.properties";

  private Phase0Runner() {}

  public static void main(String[] args) throws Exception {
    Path vectorPath = locateVector(args);
    Properties input = loadProperties(vectorPath);
    LinkedHashMap<String, String> output = calculate(input);
    emit(output);
  }

  private static LinkedHashMap<String, String> calculate(Properties input)
      throws Exception {
    byte[] compressedItems = decodeHex(required(input, "batch.compressed_items_hex"));
    byte[] batchPayloadHash = sha256(compressedItems);
    requireEqualHex(
        "batch.payload_sha256_hex",
        required(input, "batch.payload_sha256_hex"),
        batchPayloadHash);

    byte[] lcb1 = buildLcb1(input, batchPayloadHash);
    byte[] lce1 = buildLce1(input);
    byte[] submittedHash = sha256(lce1);

    String canonicalUserId = required(input, "kafka_key_fields.0");
    byte[] lcc1 = buildIdentityFrame("LCC1", canonicalUserId, submittedHash);
    byte[] canonicalHash = sha256(lcc1);

    byte[] rawSeriesPayload = decodeHex(required(input, "series.raw_payload_hex"));
    byte[] seriesChecksum = sha256(rawSeriesPayload);
    requireEqualHex(
        "series.checksum_hex",
        required(input, "series.checksum_hex"),
        seriesChecksum);

    byte[] submittedSeriesWire =
        decodeHex(required(input, "series.submitted_wire_hex"));
    byte[] lcs1 = buildFramedPayload("LCS1", submittedSeriesWire);
    byte[] seriesSubmittedHash = sha256(lcs1);
    byte[] lcr1 =
        buildIdentityFrame(
            "LCR1",
            required(input, "series.user_id"),
            seriesSubmittedHash);
    byte[] seriesCanonicalHash = sha256(lcr1);

    byte[] compressedSeries =
        decodeHex(required(input, "series.compressed_payload_hex"));
    byte[] compressedHash = sha256(compressedSeries);
    String objectVersion = encodeHex(compressedHash);
    String objectKey =
        "private/"
            + required(input, "series.user_id")
            + "/series/"
            + required(input, "series.chunk_id")
            + "/"
            + objectVersion
            + ".zst";

    byte[] kafkaKey = buildKafkaKey(input);

    LinkedHashMap<String, String> output = new LinkedHashMap<>();
    output.put("language", "java");
    output.put("lcb1_hex", encodeHex(lcb1));
    output.put("lce1_hex", encodeHex(lce1));
    output.put("lcc1_hex", encodeHex(lcc1));
    output.put("lcs1_hex", encodeHex(lcs1));
    output.put("lcr1_hex", encodeHex(lcr1));
    output.put("submitted_sha256_hex", encodeHex(submittedHash));
    output.put("canonical_sha256_hex", encodeHex(canonicalHash));
    output.put(
        "series_submitted_sha256_hex", encodeHex(seriesSubmittedHash));
    output.put(
        "series_canonical_sha256_hex", encodeHex(seriesCanonicalHash));
    output.put("compressed_sha256_hex", objectVersion);
    output.put("kafka_key_hex", encodeHex(kafkaKey));
    output.put("compressed_size", Integer.toString(compressedSeries.length));
    output.put("object_version", objectVersion);
    output.put("object_key", objectKey);

    decodeEvent(input, output);
    return output;
  }

  private static byte[] buildLcb1(Properties input, byte[] payloadHash) {
    ByteArrayOutputStream output = new ByteArrayOutputStream();
    appendAscii(output, "LCB1");
    appendFramedString(output, required(input, "batch.batch_id"));
    appendFramedString(output, required(input, "batch.device_id"));
    appendFramedString(
        output, required(input, "batch.collector_instance_id"));
    appendU64(output, parseU64(input, "batch.sequence_start"));
    appendU64(output, parseU64(input, "batch.sequence_end"));
    appendI64(output, parseI64(input, "batch.created_at.seconds"));
    appendU32(output, parseU32(input, "batch.created_at.nanos"));
    appendFramedBytes(output, decodeHex(required(input, "batch.nonce_hex")));
    appendU32(output, parseU32(input, "batch.compression"));
    appendFramedString(output, required(input, "batch.source"));
    append(output, payloadHash);
    return output.toByteArray();
  }

  private static byte[] buildLce1(Properties input) {
    ByteArrayOutputStream output = new ByteArrayOutputStream();
    appendAscii(output, "LCE1");
    appendFramedString(output, required(input, "event.event_id"));
    appendFramedString(output, required(input, "event.stream"));
    appendFramedString(output, required(input, "event.event_type"));
    appendU32(output, parseU32(input, "event.kind"));
    appendFramedString(output, required(input, "event.device_id"));
    appendFramedString(
        output, required(input, "event.collector_instance_id"));
    appendFramedString(output, required(input, "event.source"));
    appendU32(output, parseU32(input, "event.schema_version"));
    appendU64(output, parseU64(input, "event.sequence"));
    appendI64(output, parseI64(input, "event.observed_at.seconds"));
    appendU32(output, parseU32(input, "event.observed_at.nanos"));

    boolean endedAtPresent =
        parseBoolean(input, "event.ended_at.present");
    output.write(endedAtPresent ? 1 : 0);
    if (endedAtPresent) {
      appendI64(output, parseI64(input, "event.ended_at.seconds"));
      appendU32(output, parseU32(input, "event.ended_at.nanos"));
    }

    appendFramedString(output, required(input, "event.timezone"));
    appendU32(output, parseU32(input, "event.privacy_class"));
    appendU32(output, parseU32(input, "event.retention_class"));
    appendFramedString(output, required(input, "event.origin.provider"));
    appendFramedString(
        output, required(input, "event.origin.provider_record_id"));
    appendFramedString(output, required(input, "event.origin.import_id"));
    appendFramedString(
        output, required(input, "event.origin.parent_event_id"));
    appendFramedString(
        output, required(input, "event.origin.collection_method"));
    appendFramedString(output, required(input, "event.payload_type_url"));
    appendFramedBytes(
        output, decodeHex(required(input, "event.payload_value_hex")));
    return output.toByteArray();
  }

  private static byte[] buildIdentityFrame(
      String magic, String userId, byte[] submittedHash) {
    ByteArrayOutputStream output = new ByteArrayOutputStream();
    appendAscii(output, magic);
    appendFramedString(output, userId);
    append(output, submittedHash);
    return output.toByteArray();
  }

  private static byte[] buildFramedPayload(String magic, byte[] payload) {
    ByteArrayOutputStream output = new ByteArrayOutputStream();
    appendAscii(output, magic);
    appendFramedBytes(output, payload);
    return output.toByteArray();
  }

  private static byte[] buildKafkaKey(Properties input) {
    int count = parseNonNegativeInt(input, "kafka_key_fields.count");
    if (count == 0) {
      throw new IllegalArgumentException(
          "kafka_key_fields.count must be greater than zero");
    }
    ByteArrayOutputStream output = new ByteArrayOutputStream();
    for (int index = 0; index < count; index++) {
      appendFramedString(
          output, required(input, "kafka_key_fields." + index));
    }
    return output.toByteArray();
  }

  private static void decodeEvent(
      Properties input, LinkedHashMap<String, String> output)
      throws Exception {
    byte[] wire = decodeHex(required(input, "protobuf_event_hex"));
    EventEnvelope event = EventEnvelope.parseFrom(wire);
    if (!event.hasObservedAt()) {
      throw new IllegalArgumentException(
          "protobuf_event_hex has no observed_at");
    }
    if (!event.hasOrigin()) {
      throw new IllegalArgumentException("protobuf_event_hex has no origin");
    }
    if (!event.hasPayload()) {
      throw new IllegalArgumentException("protobuf_event_hex has no payload");
    }

    Any anyPayload = event.getPayload();
    if (!anyPayload.is(AppForeground.class)) {
      throw new IllegalArgumentException(
          "protobuf_event_hex payload is not AppForeground: "
              + anyPayload.getTypeUrl());
    }
    AppForeground payload = anyPayload.unpack(AppForeground.class);
    Origin origin = event.getOrigin();

    output.put("decoded.event_id", event.getEventId());
    output.put("decoded.stream", event.getStream());
    output.put("decoded.event_type", event.getEventType());
    output.put("decoded.kind", Integer.toString(event.getKindValue()));
    output.put("decoded.user_id", event.getUserId());
    output.put("decoded.device_id", event.getDeviceId());
    output.put(
        "decoded.collector_instance_id", event.getCollectorInstanceId());
    output.put("decoded.source", event.getSource());
    output.put(
        "decoded.schema_version",
        Integer.toUnsignedString(event.getSchemaVersion()));
    output.put(
        "decoded.sequence", Long.toUnsignedString(event.getSequence()));
    output.put(
        "decoded.observed_at.seconds",
        Long.toString(event.getObservedAt().getSeconds()));
    output.put(
        "decoded.observed_at.nanos",
        Integer.toString(event.getObservedAt().getNanos()));
    output.put("decoded.ended_at.present", Boolean.toString(event.hasEndedAt()));
    output.put("decoded.timezone", event.getTimezone());
    output.put(
        "decoded.privacy_class",
        Integer.toString(event.getPrivacyClassValue()));
    output.put(
        "decoded.retention_class",
        Integer.toString(event.getRetentionClassValue()));
    output.put("decoded.origin.provider", origin.getProvider());
    output.put(
        "decoded.origin.provider_record_id", origin.getProviderRecordId());
    output.put("decoded.origin.import_id", origin.getImportId());
    output.put(
        "decoded.origin.parent_event_id", origin.getParentEventId());
    output.put(
        "decoded.origin.collection_method", origin.getCollectionMethod());
    output.put("decoded.payload_type_url", anyPayload.getTypeUrl());
    output.put(
        "decoded.payload_value_hex",
        encodeHex(anyPayload.getValue().toByteArray()));
    output.put(
        "decoded.payload.application_id", payload.getApplicationId());
    output.put(
        "decoded.payload.application_name", payload.getApplicationName());
    output.put(
        "decoded.payload.executable_name", payload.getExecutableName());
    output.put(
        "decoded.payload.window_title_utf8_hex",
        encodeHex(payload.getWindowTitle().getBytes(StandardCharsets.UTF_8)));
    output.put(
        "decoded.payload.process_id",
        Long.toUnsignedString(payload.getProcessId()));
    output.put("decoded.payload.window_id", payload.getWindowId());
    output.put(
        "decoded.payload.fullscreen",
        Boolean.toString(payload.getFullscreen()));
  }

  private static Properties loadProperties(Path path) throws IOException {
    Properties properties = new Properties();
    try (var reader = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
      properties.load(reader);
    }
    return properties;
  }

  private static Path locateVector(String[] args) {
    if (args.length > 1) {
      throw new IllegalArgumentException(
          "usage: Phase0Runner [path/to/" + VECTOR_NAME + "]");
    }
    if (args.length == 1) {
      Path explicit = Path.of(args[0]).toAbsolutePath().normalize();
      if (!Files.isRegularFile(explicit)) {
        throw new IllegalArgumentException(
            "vector file does not exist: " + explicit);
      }
      return explicit;
    }

    Path cursor = Path.of("").toAbsolutePath().normalize();
    while (cursor != null) {
      Path repositoryCandidate =
          cursor.resolve("contract-tests").resolve("vectors").resolve(VECTOR_NAME);
      if (Files.isRegularFile(repositoryCandidate)) {
        return repositoryCandidate;
      }
      Path contractTestsCandidate =
          cursor.resolve("vectors").resolve(VECTOR_NAME);
      if (Files.isRegularFile(contractTestsCandidate)) {
        return contractTestsCandidate;
      }
      cursor = cursor.getParent();
    }
    throw new IllegalArgumentException(
        "cannot locate contract-tests/vectors/" + VECTOR_NAME);
  }

  private static String required(Properties input, String key) {
    if (key.startsWith("expected.")) {
      throw new IllegalArgumentException(
          "runner code must not read expected.* properties: " + key);
    }
    String value = input.getProperty(key);
    if (value == null) {
      throw new IllegalArgumentException("missing vector property: " + key);
    }
    return value;
  }

  private static int parseNonNegativeInt(Properties input, String key) {
    String raw = required(input, key);
    try {
      int value = Integer.parseInt(raw);
      if (value < 0) {
        throw new IllegalArgumentException(
            key + " must be non-negative: " + raw);
      }
      return value;
    } catch (NumberFormatException exc) {
      throw new IllegalArgumentException(key + " is not an integer: " + raw, exc);
    }
  }

  private static long parseU32(Properties input, String key) {
    String raw = required(input, key);
    try {
      long value = Long.parseLong(raw);
      if (value < 0 || value > 0xffff_ffffL) {
        throw new IllegalArgumentException(
            key + " is outside uint32: " + raw);
      }
      return value;
    } catch (NumberFormatException exc) {
      throw new IllegalArgumentException(key + " is not uint32: " + raw, exc);
    }
  }

  private static long parseU64(Properties input, String key) {
    String raw = required(input, key);
    try {
      return Long.parseUnsignedLong(raw);
    } catch (NumberFormatException exc) {
      throw new IllegalArgumentException(key + " is not uint64: " + raw, exc);
    }
  }

  private static long parseI64(Properties input, String key) {
    String raw = required(input, key);
    try {
      return Long.parseLong(raw);
    } catch (NumberFormatException exc) {
      throw new IllegalArgumentException(key + " is not int64: " + raw, exc);
    }
  }

  private static boolean parseBoolean(Properties input, String key) {
    String raw = required(input, key);
    if (raw.equals("true")) {
      return true;
    }
    if (raw.equals("false")) {
      return false;
    }
    throw new IllegalArgumentException(key + " is not true or false: " + raw);
  }

  private static void requireEqualHex(
      String key, String declaredHex, byte[] calculated) {
    String normalized = declaredHex.toLowerCase(Locale.ROOT);
    String actual = encodeHex(calculated);
    if (!normalized.equals(actual)) {
      throw new IllegalArgumentException(
          key + " does not match independently calculated SHA-256");
    }
  }

  private static byte[] sha256(byte[] value) {
    try {
      return MessageDigest.getInstance("SHA-256").digest(value);
    } catch (NoSuchAlgorithmException exc) {
      throw new AssertionError("JVM has no SHA-256 provider", exc);
    }
  }

  private static byte[] decodeHex(String value) {
    if ((value.length() & 1) != 0) {
      throw new IllegalArgumentException("hex value has odd length");
    }
    byte[] output = new byte[value.length() / 2];
    for (int index = 0; index < output.length; index++) {
      int high = Character.digit(value.charAt(index * 2), 16);
      int low = Character.digit(value.charAt(index * 2 + 1), 16);
      if (high < 0 || low < 0) {
        throw new IllegalArgumentException(
            "hex value contains a non-hexadecimal character");
      }
      output[index] = (byte) ((high << 4) | low);
    }
    return output;
  }

  private static String encodeHex(byte[] value) {
    char[] digits = "0123456789abcdef".toCharArray();
    char[] output = new char[value.length * 2];
    for (int index = 0; index < value.length; index++) {
      int unsigned = value[index] & 0xff;
      output[index * 2] = digits[unsigned >>> 4];
      output[index * 2 + 1] = digits[unsigned & 0x0f];
    }
    return new String(output);
  }

  private static void appendAscii(ByteArrayOutputStream output, String value) {
    append(output, value.getBytes(StandardCharsets.US_ASCII));
  }

  private static void appendFramedString(
      ByteArrayOutputStream output, String value) {
    appendFramedBytes(output, value.getBytes(StandardCharsets.UTF_8));
  }

  private static void appendFramedBytes(
      ByteArrayOutputStream output, byte[] value) {
    appendU32(output, value.length);
    append(output, value);
  }

  private static void appendU32(ByteArrayOutputStream output, long value) {
    if (value < 0 || value > 0xffff_ffffL) {
      throw new IllegalArgumentException("value is outside uint32: " + value);
    }
    output.write((int) (value >>> 24) & 0xff);
    output.write((int) (value >>> 16) & 0xff);
    output.write((int) (value >>> 8) & 0xff);
    output.write((int) value & 0xff);
  }

  private static void appendU64(ByteArrayOutputStream output, long value) {
    appendI64(output, value);
  }

  private static void appendI64(ByteArrayOutputStream output, long value) {
    for (int shift = 56; shift >= 0; shift -= 8) {
      output.write((int) (value >>> shift) & 0xff);
    }
  }

  private static void append(ByteArrayOutputStream output, byte[] value) {
    output.writeBytes(value);
  }

  private static void emit(LinkedHashMap<String, String> output) {
    try (PrintWriter writer =
        new PrintWriter(
            new OutputStreamWriter(System.out, StandardCharsets.UTF_8), false)) {
      for (Map.Entry<String, String> entry : output.entrySet()) {
        String key = entry.getKey();
        String value = entry.getValue();
        if (key.indexOf('=') >= 0
            || key.indexOf('\n') >= 0
            || key.indexOf('\r') >= 0
            || value.indexOf('\n') >= 0
            || value.indexOf('\r') >= 0) {
          throw new IllegalArgumentException(
              "stdout key/value contains an unsupported delimiter");
        }
        writer.print(key);
        writer.print('=');
        writer.println(value);
      }
      writer.flush();
    }
  }
}

File: contract-tests/runners/kotlin/Phase0Runner.kt
import com.google.protobuf.Any
import io.lifechronicle.events.v1.AppForeground
import io.lifechronicle.events.v1.EventEnvelope
import io.lifechronicle.events.v1.copy
import java.io.ByteArrayOutputStream
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.security.MessageDigest
import java.util.Locale
import java.util.Properties

/**
 * Independent Kotlin implementation of the Phase 0 golden-vector framing.
 *
 * The generated Kotlin API augments the generated Java message classes, so the
 * no-op copy DSL calls below intentionally force both generated source sets to
 * compile and link while parseFrom performs the Protobuf wire decode.
 */
private const val VECTOR_NAME = "phase0-v1.properties"

fun main(args: Array<String>) {
    val vectorPath = locateVector(args)
    val input = loadProperties(vectorPath)
    emit(calculate(input))
}

private fun calculate(input: Properties): LinkedHashMap<String, String> {
    val compressedItems = decodeHex(required(input, "batch.compressed_items_hex"))
    val batchPayloadHash = sha256(compressedItems)
    requireEqualHex(
        "batch.payload_sha256_hex",
        required(input, "batch.payload_sha256_hex"),
        batchPayloadHash,
    )

    val lcb1 = buildLcb1(input, batchPayloadHash)
    val lce1 = buildLce1(input)
    val submittedHash = sha256(lce1)

    val canonicalUserId = required(input, "kafka_key_fields.0")
    val lcc1 = buildIdentityFrame("LCC1", canonicalUserId, submittedHash)
    val canonicalHash = sha256(lcc1)

    val rawSeriesPayload = decodeHex(required(input, "series.raw_payload_hex"))
    val seriesChecksum = sha256(rawSeriesPayload)
    requireEqualHex(
        "series.checksum_hex",
        required(input, "series.checksum_hex"),
        seriesChecksum,
    )

    val submittedSeriesWire =
        decodeHex(required(input, "series.submitted_wire_hex"))
    val lcs1 = buildFramedPayload("LCS1", submittedSeriesWire)
    val seriesSubmittedHash = sha256(lcs1)
    val lcr1 =
        buildIdentityFrame(
            "LCR1",
            required(input, "series.user_id"),
            seriesSubmittedHash,
        )
    val seriesCanonicalHash = sha256(lcr1)

    val compressedSeries =
        decodeHex(required(input, "series.compressed_payload_hex"))
    val compressedHash = sha256(compressedSeries)
    val objectVersion = encodeHex(compressedHash)
    val objectKey =
        "private/${required(input, "series.user_id")}/series/" +
            "${required(input, "series.chunk_id")}/$objectVersion.zst"
    val kafkaKey = buildKafkaKey(input)

    val output = linkedMapOf<String, String>()
    output["language"] = "kotlin"
    output["lcb1_hex"] = encodeHex(lcb1)
    output["lce1_hex"] = encodeHex(lce1)
    output["lcc1_hex"] = encodeHex(lcc1)
    output["lcs1_hex"] = encodeHex(lcs1)
    output["lcr1_hex"] = encodeHex(lcr1)
    output["submitted_sha256_hex"] = encodeHex(submittedHash)
    output["canonical_sha256_hex"] = encodeHex(canonicalHash)
    output["series_submitted_sha256_hex"] = encodeHex(seriesSubmittedHash)
    output["series_canonical_sha256_hex"] = encodeHex(seriesCanonicalHash)
    output["compressed_sha256_hex"] = objectVersion
    output["kafka_key_hex"] = encodeHex(kafkaKey)
    output["compressed_size"] = compressedSeries.size.toString()
    output["object_version"] = objectVersion
    output["object_key"] = objectKey

    decodeEvent(input, output)
    return output
}

private fun buildLcb1(input: Properties, payloadHash: ByteArray): ByteArray =
    ByteArrayOutputStream().run {
        appendAscii("LCB1")
        appendFramedString(required(input, "batch.batch_id"))
        appendFramedString(required(input, "batch.device_id"))
        appendFramedString(required(input, "batch.collector_instance_id"))
        appendU64(parseU64(input, "batch.sequence_start"))
        appendU64(parseU64(input, "batch.sequence_end"))
        appendI64(parseI64(input, "batch.created_at.seconds"))
        appendU32(parseU32(input, "batch.created_at.nanos"))
        appendFramedBytes(decodeHex(required(input, "batch.nonce_hex")))
        appendU32(parseU32(input, "batch.compression"))
        appendFramedString(required(input, "batch.source"))
        write(payloadHash)
        toByteArray()
    }

private fun buildLce1(input: Properties): ByteArray =
    ByteArrayOutputStream().run {
        appendAscii("LCE1")
        appendFramedString(required(input, "event.event_id"))
        appendFramedString(required(input, "event.stream"))
        appendFramedString(required(input, "event.event_type"))
        appendU32(parseU32(input, "event.kind"))
        appendFramedString(required(input, "event.device_id"))
        appendFramedString(required(input, "event.collector_instance_id"))
        appendFramedString(required(input, "event.source"))
        appendU32(parseU32(input, "event.schema_version"))
        appendU64(parseU64(input, "event.sequence"))
        appendI64(parseI64(input, "event.observed_at.seconds"))
        appendU32(parseU32(input, "event.observed_at.nanos"))

        val endedAtPresent = parseBoolean(input, "event.ended_at.present")
        write(if (endedAtPresent) 1 else 0)
        if (endedAtPresent) {
            appendI64(parseI64(input, "event.ended_at.seconds"))
            appendU32(parseU32(input, "event.ended_at.nanos"))
        }

        appendFramedString(required(input, "event.timezone"))
        appendU32(parseU32(input, "event.privacy_class"))
        appendU32(parseU32(input, "event.retention_class"))
        appendFramedString(required(input, "event.origin.provider"))
        appendFramedString(required(input, "event.origin.provider_record_id"))
        appendFramedString(required(input, "event.origin.import_id"))
        appendFramedString(required(input, "event.origin.parent_event_id"))
        appendFramedString(required(input, "event.origin.collection_method"))
        appendFramedString(required(input, "event.payload_type_url"))
        appendFramedBytes(decodeHex(required(input, "event.payload_value_hex")))
        toByteArray()
    }

private fun buildIdentityFrame(
    magic: String,
    userId: String,
    submittedHash: ByteArray,
): ByteArray =
    ByteArrayOutputStream().run {
        appendAscii(magic)
        appendFramedString(userId)
        write(submittedHash)
        toByteArray()
    }

private fun buildFramedPayload(magic: String, payload: ByteArray): ByteArray =
    ByteArrayOutputStream().run {
        appendAscii(magic)
        appendFramedBytes(payload)
        toByteArray()
    }

private fun buildKafkaKey(input: Properties): ByteArray {
    val count = parseNonNegativeInt(input, "kafka_key_fields.count")
    require(count > 0) { "kafka_key_fields.count must be greater than zero" }
    return ByteArrayOutputStream().run {
        repeat(count) { index ->
            appendFramedString(required(input, "kafka_key_fields.$index"))
        }
        toByteArray()
    }
}

private fun decodeEvent(
    input: Properties,
    output: LinkedHashMap<String, String>,
) {
    val wire = decodeHex(required(input, "protobuf_event_hex"))
    val event =
        EventEnvelope.parseFrom(wire).copy {
            // Generated Kotlin copy DSL intentionally participates in decode.
        }
    require(event.hasObservedAt()) { "protobuf_event_hex has no observed_at" }
    require(event.hasOrigin()) { "protobuf_event_hex has no origin" }
    require(event.hasPayload()) { "protobuf_event_hex has no payload" }

    val anyPayload: Any = event.payload
    require(anyPayload.`is`(AppForeground::class.java)) {
        "protobuf_event_hex payload is not AppForeground: ${anyPayload.typeUrl}"
    }
    val payload =
        anyPayload.unpack(AppForeground::class.java).copy {
            // Generated Kotlin copy DSL intentionally participates in decode.
        }
    val origin = event.origin

    output["decoded.event_id"] = event.eventId
    output["decoded.stream"] = event.stream
    output["decoded.event_type"] = event.eventType
    output["decoded.kind"] = event.kindValue.toString()
    output["decoded.user_id"] = event.userId
    output["decoded.device_id"] = event.deviceId
    output["decoded.collector_instance_id"] = event.collectorInstanceId
    output["decoded.source"] = event.source
    output["decoded.schema_version"] =
        Integer.toUnsignedString(event.schemaVersion)
    output["decoded.sequence"] = java.lang.Long.toUnsignedString(event.sequence)
    output["decoded.observed_at.seconds"] = event.observedAt.seconds.toString()
    output["decoded.observed_at.nanos"] = event.observedAt.nanos.toString()
    output["decoded.ended_at.present"] = event.hasEndedAt().toString()
    output["decoded.timezone"] = event.timezone
    output["decoded.privacy_class"] = event.privacyClassValue.toString()
    output["decoded.retention_class"] = event.retentionClassValue.toString()
    output["decoded.origin.provider"] = origin.provider
    output["decoded.origin.provider_record_id"] = origin.providerRecordId
    output["decoded.origin.import_id"] = origin.importId
    output["decoded.origin.parent_event_id"] = origin.parentEventId
    output["decoded.origin.collection_method"] = origin.collectionMethod
    output["decoded.payload_type_url"] = anyPayload.typeUrl
    output["decoded.payload_value_hex"] =
        encodeHex(anyPayload.value.toByteArray())
    output["decoded.payload.application_id"] = payload.applicationId
    output["decoded.payload.application_name"] = payload.applicationName
    output["decoded.payload.executable_name"] = payload.executableName
    output["decoded.payload.window_title_utf8_hex"] =
        encodeHex(payload.windowTitle.toByteArray(StandardCharsets.UTF_8))
    output["decoded.payload.process_id"] =
        java.lang.Long.toUnsignedString(payload.processId)
    output["decoded.payload.window_id"] = payload.windowId
    output["decoded.payload.fullscreen"] = payload.fullscreen.toString()
}

private fun loadProperties(path: Path): Properties =
    Properties().also { properties ->
        Files.newBufferedReader(path, StandardCharsets.UTF_8).use {
            properties.load(it)
        }
    }

private fun locateVector(args: Array<String>): Path {
    require(args.size <= 1) { "usage: Phase0RunnerKt [path/to/$VECTOR_NAME]" }
    if (args.size == 1) {
        val explicit = Path.of(args[0]).toAbsolutePath().normalize()
        require(Files.isRegularFile(explicit)) {
            "vector file does not exist: $explicit"
        }
        return explicit
    }

    var cursor: Path? = Path.of("").toAbsolutePath().normalize()
    while (cursor != null) {
        val repositoryCandidate =
            cursor.resolve("contract-tests").resolve("vectors").resolve(VECTOR_NAME)
        if (Files.isRegularFile(repositoryCandidate)) {
            return repositoryCandidate
        }
        val contractTestsCandidate = cursor.resolve("vectors").resolve(VECTOR_NAME)
        if (Files.isRegularFile(contractTestsCandidate)) {
            return contractTestsCandidate
        }
        cursor = cursor.parent
    }
    throw IllegalArgumentException(
        "cannot locate contract-tests/vectors/$VECTOR_NAME",
    )
}

private fun required(input: Properties, key: String): String {
    require(!key.startsWith("expected.")) {
        "runner code must not read expected.* properties: $key"
    }
    return input.getProperty(key)
        ?: throw IllegalArgumentException("missing vector property: $key")
}

private fun parseNonNegativeInt(input: Properties, key: String): Int {
    val raw = required(input, key)
    val value =
        raw.toIntOrNull()
            ?: throw IllegalArgumentException("$key is not an integer: $raw")
    require(value >= 0) { "$key must be non-negative: $raw" }
    return value
}

private fun parseU32(input: Properties, key: String): Long {
    val raw = required(input, key)
    val value =
        raw.toLongOrNull()
            ?: throw IllegalArgumentException("$key is not uint32: $raw")
    require(value in 0..0xffff_ffffL) { "$key is outside uint32: $raw" }
    return value
}

private fun parseU64(input: Properties, key: String): Long {
    val raw = required(input, key)
    return try {
        java.lang.Long.parseUnsignedLong(raw)
    } catch (exception: NumberFormatException) {
        throw IllegalArgumentException("$key is not uint64: $raw", exception)
    }
}

private fun parseI64(input: Properties, key: String): Long {
    val raw = required(input, key)
    return raw.toLongOrNull()
        ?: throw IllegalArgumentException("$key is not int64: $raw")
}

private fun parseBoolean(input: Properties, key: String): Boolean =
    when (val raw = required(input, key)) {
        "true" -> true
        "false" -> false
        else -> throw IllegalArgumentException(
            "$key is not true or false: $raw",
        )
    }

private fun requireEqualHex(
    key: String,
    declaredHex: String,
    calculated: ByteArray,
) {
    require(declaredHex.lowercase(Locale.ROOT) == encodeHex(calculated)) {
        "$key does not match independently calculated SHA-256"
    }
}

private fun sha256(value: ByteArray): ByteArray =
    MessageDigest.getInstance("SHA-256").digest(value)

private fun decodeHex(value: String): ByteArray {
    require(value.length % 2 == 0) { "hex value has odd length" }
    return ByteArray(value.length / 2) { index ->
        val high = Character.digit(value[index * 2], 16)
        val low = Character.digit(value[index * 2 + 1], 16)
        require(high >= 0 && low >= 0) {
            "hex value contains a non-hexadecimal character"
        }
        ((high shl 4) or low).toByte()
    }
}

private fun encodeHex(value: ByteArray): String {
    val digits = "0123456789abcdef"
    return buildString(value.size * 2) {
        value.forEach { byte ->
            val unsigned = byte.toInt() and 0xff
            append(digits[unsigned ushr 4])
            append(digits[unsigned and 0x0f])
        }
    }
}

private fun ByteArrayOutputStream.appendAscii(value: String) {
    write(value.toByteArray(StandardCharsets.US_ASCII))
}

private fun ByteArrayOutputStream.appendFramedString(value: String) {
    appendFramedBytes(value.toByteArray(StandardCharsets.UTF_8))
}

private fun ByteArrayOutputStream.appendFramedBytes(value: ByteArray) {
    appendU32(value.size.toLong())
    write(value)
}

private fun ByteArrayOutputStream.appendU32(value: Long) {
    require(value in 0..0xffff_ffffL) { "value is outside uint32: $value" }
    write(((value ushr 24) and 0xff).toInt())
    write(((value ushr 16) and 0xff).toInt())
    write(((value ushr 8) and 0xff).toInt())
    write((value and 0xff).toInt())
}

private fun ByteArrayOutputStream.appendU64(value: Long) {
    appendI64(value)
}

private fun ByteArrayOutputStream.appendI64(value: Long) {
    for (shift in 56 downTo 0 step 8) {
        write(((value ushr shift) and 0xff).toInt())
    }
}

private fun emit(output: LinkedHashMap<String, String>) {
    val writer = OutputStreamWriter(System.out, StandardCharsets.UTF_8)
    output.forEach { (key, value) ->
        require('=' !in key && '\n' !in key && '\r' !in key) {
            "stdout key contains an unsupported delimiter"
        }
        require('\n' !in value && '\r' !in value) {
            "stdout value contains an unsupported delimiter"
        }
        writer.write(key)
        writer.write("=")
        writer.write(value)
        writer.write("\n")
    }
    writer.flush()
}
